File: android/Example/app/src/main/java/sdk/fluig/com/example/main/presenter/MainPresenter.java
package sdk.fluig.com.example.main.presenter;

import sdk.fluig.com.example.R;
import sdk.fluig.com.example.main.contract.MainContract;
import sdk.fluig.com.example.model.ComponentType;
import sdk.fluig.com.example.model.FlowType;
import sdk.fluig.com.example.model.ItemType;
import sdk.fluig.com.example.model.ListType;

/**
 * Created by gregorysholl on 26/02/18.
 */

public class MainPresenter implements MainContract.Presenter {

    private MainContract.View mView;

    public MainPresenter(MainContract.View view) {
        mView = view;
        mView.setPresenter(this);
    }

    @Override
    public void obtainNextStep(ItemType itemType) {
        if (itemType instanceof ListType) {
            showList((ListType) itemType);

        } else if (itemType instanceof FlowType) {
            showFlow((FlowType) itemType);

        } else if (itemType instanceof ComponentType) {
            mView.showComponent((ComponentType) itemType);

        } else {
            mView.showError(R.string.list_notfound);
        }
    }

    private void showList(ListType listType) {
        switch (listType) {
            case FLOW:
                mView.showList(FlowType.values());
                break;
            case COMPONENTS:
                mView.showList(ComponentType.values());
                break;
        }
    }

    private void showFlow(FlowType flowType) {
        switch (flowType) {
            case LOGIN:
                mView.showLoginFlow();
                break;
            case EULA:
                mView.showEulaFlow();
                break;
        }
    }
}

File: ios/Example/Example/Controllers/Main/MainPresenter.swift
//
//  MainPresenter.swift
//  Example
//
//  Created by Gregory Sholl e Santos on 05/03/18.
//  Copyright © 2018 fluig SDK. All rights reserved.
//

import Foundation

protocol MainPresenterDelegate {
    func obtainNextStep(with itemType: ItemType)
}

class MainPresenter {
    
    private weak var view: MainViewDelegate?
    
    init(view: MainViewDelegate) {
        self.view = view
    }
    
    private func showList(_ listType: ListType) {
        switch listType {
        case .flows:
            view?.showList(FlowType.allValues)
        case .components:
            view?.showList(ComponentType.allValues)
        }
    }
    
    private func showFlow(_ flowType: FlowType) {
        switch flowType {
        case .login:
            view?.showLoginFlow()
        case .eula:
            view?.showEulaFlow()
        }
    }
}

extension MainPresenter: MainPresenterDelegate {
    
    func obtainNextStep(with itemType: ItemType) {
        if let listType = itemType as? ListType {
            showList(listType)
            
        } else if let flowType = itemType as? FlowType {
            showFlow(flowType)
            
        } else if let componentType = itemType as? ComponentType {
            view?.showComponent(componentType)
            
        } else {
            view?.showError("")
        }
    }
}
